# Assignment 1

#### Link for github 
https://github.com/murrodroid/CompSocSci

#### Contributions
Everything has been made in close collaboration. If we were forced to divide it up at gunpoint, it would be the following:
1. Mads
2. Shared
3. Christian
4. Lucas

## Part 1

Import libraries for part 1 of the assignment

In [1]:
from bs4 import BeautifulSoup
import requests
import csv

In [2]:
link = "https://ic2s2-2023.org/program"
req = requests.get(link)
soup = BeautifulSoup(req.content)

Extracting names with i meaning chairs, parallel talks and authors of posters

In [3]:
table_rows = soup.find_all("i")
names = []
for line in table_rows:
    line = line.get_text(strip=True).replace(":", ",").split(",")
    if "Chair" in line:
        line.remove("Chair")
    names.extend(name.strip() for name in line if name.strip())
names = list(dict.fromkeys(names))

print((names))

print(len(names))

['Claudia Wagner', 'Jonas L Juul', 'Jon Kleinberg', 'Chloe Ahn', 'Xinyi Wang', 'Giuseppe Russo', 'luca verginer', 'Manoel Horta Ribeiro', 'Giona Casiraghi', 'Almog Simchon', 'Adam Sutton', 'Matthew Edwards', 'Stephan Lewandowsky', 'Arianna Pera', 'Manuel Vimercati', 'Matteo Palmonari', 'Mohammed Alsobay', 'Abdullah Almaatouq', 'David G. Rand', 'Duncan J. Watts', 'Sara Venturini', 'Satyaki Sikdar', 'Francesco Rinaldi', 'Francesco Tudisco', 'Santo Fortunato', 'Isabella Loaiza', 'Takahiro Yabe', 'Alex Pentland', 'Taha Yasseri', 'Silvia De Sojo Caso', 'Mia Ann Jørgensen', 'Sune Lehmann', 'Laura Alessandretti', 'Emil Bakkensen Johansen', 'Mathias Wullum Nielsen', 'Rubén Rodríguez Casañ', 'Antonio Ariño Villarroya', 'Sunny Rai', 'Ashley Francisco', 'Salvatore Giorgi', 'Brenda Curtis', 'Lyle Ungar', 'Sharath Chandra Guntuku', 'Allison Koenecke', 'Eric Giannella', 'Robb Willer', 'Sharad Goel', 'Ziv Epstein', 'Hause Lin', 'Levin Brinkmann', 'Bramantyo Supriyatno', 'Iyad Rahwan', 'Sandro Ferreir

Adding keynames and removing duplicates

In [4]:
# Find all <a> tags that mention "Keynote"
keynote_links = soup.find_all("a", string=lambda text: text and "Keynote" in text)

# Extract and clean names
keynote_speakers = []
for link in keynote_links:
    text = link.get_text(strip=True)  # Get text inside <a> tag
    name = text.replace("Keynote - ", "").strip()  # Remove "Keynote - "
    keynote_speakers.append(name)
# Print all keynote speakers
if "Keynotes" in keynote_speakers:
    keynote_speakers.remove("Keynotes")

print(keynote_speakers)
names.extend(keynote_speakers)

#Remove duplicates
names = list(set(names))
names = list({name.strip().lower() for name in names})
print(names)
print(len(names))

['Jevin West', 'Linda Steg', 'Sharad Goel', 'Molly Crockett', 'Lisa Anne Hendriks', 'Stefan Gössling', 'Joanna Bryson', 'Tim Althoff', 'Lauren Brent', 'Esteban Moro']
['xinwei xu', 'erin walk', 'david mimno', 'eladio montero-porras', 'daniel matter', 'iyad rahwan', 'quinn lanners', 'marcos a. oliveira', 'giovanni mauro', 'kat albrecht', 'shatayu kulkarni', 'tomaso erseghe', 'christa brelsford', 'mariano gaston beiro', 'milena tsvetkova', 'yevgeniy golovchenko', 'anna mancini', 'tom theile', 'seongwoon kim', 'somya tomar', 'ocã©ane boulais', 'jean-philippe cointet', 'clemens m. lechner', "valeria d'andrea", 'nir grinberg', 'gianni barlacchi', 'maxim derex', 'melanie oyarzun', 'reid mcilroy-young', 'mikko kivelä', 'mathias wullum nielsen', 'omar guerrero', 'paul salmon', 'alba motes rodrigo', 'susan vermeer', 'antonin pottier', 'victor hugo orozco-olvera', 'lisa anne hendriks', 'federico barrera lemarchand', 'justyna czestochowska', 'frank berthold', 'priyanshi sharma', 'dingeman jan van

Saving authors:

In [5]:
with open("IC2S2_2023_names.txt", "w") as file:
    for name in names:
        file.write(name + "\n")

We have 1488 unique speakers.


**6. Explain the process you followed to web-scrape the page. Which choices did you make to accurately retreive as many names as possible? Which strategies did you use to assess the quality of your final list? Explain your reasoning and your choices (answer in max 150 words).**

We looped through all tags using \<a> or \<i> to find the relevant names. The \<i> tag called all authors except the keynotes which we handled afterwards. The \<a> tag called the missing keynote speakers.

BeautifulSoups find all function was used to search for the tags.

We afterwards applied as set to the list to limit the list to unique names, furthermore standardizing the names to avoid duplicates with slightly different formatting.

---

## Part 2

**1. What are pros and cons of the custom-made data used in Centola's experiment (the first study presented in the lecture) and the ready-made data used in Nicolaides's study (the second study presented in the lecture)? You can support your arguments based on the content of the lecture and the information you read in Chapter 2.3 of the book (answer in max 150 words).**

**Centolas**

The pros of Centolas experiment was the closed environment that reduced confounding variables, enhancing internal validity and enabling clear causal inference. Furthermore the test allowed for the possibility to reproduce the result again in the future.

The cons included that the artificial setting perhaps didn't match the real world, since the participants were placed in artificial networks. This could mean the that the result perhaps isn't naturally occuring. Also if the participants were aware they were in an experiment they could behave differently.

**Nicolaides**

Pros are the data comes from the real world and is generalizable, richness in the complexity and no experimental bias.

Cons are lack of control coming from real world data, data quality issues also from real world data and correlation does not imply causation so again the noisy real world data is hard to draw conclusions on.

**2. How do you think these differences can influence the interpretation of the results in each study? (answer in max 150 words)**

Centolas experiment allowed for clear control of the environment and therefore the ability to draw clear conclusions. However the data could risk not being true to the real world. Nicolaides on the other hand had more generalizable data, but lacked control of the environment making it hard to draw clear conclusion. However the data represented the real world. A combination of the two would most likely be the best option and could strengthen the reliability of scientific conclusions. 

---

## Part 3

Getting an overview:

In [6]:
import requests

def get_posts(url):
    try:
        response = requests.get(url)

        if response.status_code == 200:
            posts = response.json()
            return posts["results"]
        else:
            print('Error:', response.status_code)
            return None
    except requests.exceptions.RequestException as e:
        print('Error:', e)
        return None

posts = get_posts('https://api.openalex.org/works')
posts

[{'id': 'https://openalex.org/W1775749144',
  'doi': 'https://doi.org/10.1016/s0021-9258(19)52451-6',
  'title': 'PROTEIN MEASUREMENT WITH THE FOLIN PHENOL REAGENT',
  'display_name': 'PROTEIN MEASUREMENT WITH THE FOLIN PHENOL REAGENT',
  'publication_year': 1951,
  'publication_date': '1951-11-01',
  'ids': {'openalex': 'https://openalex.org/W1775749144',
   'doi': 'https://doi.org/10.1016/s0021-9258(19)52451-6',
   'mag': '1775749144',
   'pmid': 'https://pubmed.ncbi.nlm.nih.gov/14907713'},
  'language': 'en',
  'primary_location': {'is_oa': True,
   'landing_page_url': 'https://doi.org/10.1016/s0021-9258(19)52451-6',
   'pdf_url': None,
   'source': {'id': 'https://openalex.org/S140251998',
    'display_name': 'Journal of Biological Chemistry',
    'issn_l': '0021-9258',
    'issn': ['0021-9258', '1067-8816', '1083-351X'],
    'is_oa': True,
    'is_in_doaj': True,
    'is_indexed_in_scopus': True,
    'is_core': True,
    'host_organization': 'https://openalex.org/P4310320990',
   

Importing names to work with:

In [7]:
import pandas as pd

# Load Names and Universities from previous weeks exercise
df = pd.read_csv("authors_matched.csv")
Names = list(df["Name"])
id_author = list(df["ID"])

Since we have a relevant name in the first post from the API, lets see if we can get a hold of the correct data items:

In [8]:
# Process the collected posts
ids = []
publication_year = []
cited_by_count = []
author_ids = []
titles = []
abstracts_inverted_index = []
for i in posts:
    for j in i["authorships"]:
        if (j["author"]["display_name"]) in Names:
            # print(j["author"]["display_name"],j["author"]["id"],i["id"],i["publication_year"],i["cited_by_count"],i["title"],i["abstract_inverted_index"])
            ids.append(i["id"])
            publication_year.append(i["publication_year"])
            cited_by_count.append(i["cited_by_count"])
            titles.append(i["title"])
            abstracts_inverted_index.append(i["abstract_inverted_index"])
            x = []
            for aid in i["authorships"]: 
                x.append(aid["author"]["id"])
            author_ids.append([", ".join(x)])
print(ids,"\n",len(ids),"\n",publication_year,"\n",len(publication_year),"\n",cited_by_count,"\n",len(cited_by_count),"\n",author_ids,"\n",len(author_ids),"\n",titles,"\n",len(titles),"\n",abstracts_inverted_index,"\n",len(abstracts_inverted_index))

['https://openalex.org/W2194775991'] 
 1 
 [2016] 
 1 
 [184257] 
 1 
 [['https://openalex.org/A5100700361, https://openalex.org/A5100362465, https://openalex.org/A5100824251, https://openalex.org/A5100785015']] 
 1 
 ['Deep Residual Learning for Image Recognition'] 
 1 
 [{'Deeper': [0], 'neural': [1], 'networks': [2, 19, 57], 'are': [3, 21, 58, 170], 'more': [4], 'difficult': [5], 'to': [6, 14, 39, 60, 83, 150, 175], 'train.': [7], 'We': [8, 28, 48, 124], 'present': [9, 126], 'a': [10, 79, 157], 'residual': [11, 35, 56, 76, 102, 168], 'learning': [12, 34, 45], 'framework': [13], 'ease': [15], 'the': [16, 31, 40, 71, 108, 115, 119, 162, 185, 189], 'training': [17], 'of': [18, 44, 81, 100, 137, 140, 172, 191], 'that': [20, 54], 'substantially': [22], 'deeper': [23, 88], 'than': [24, 89], 'those': [25], 'used': [26], 'previously.': [27], 'explicitly': [29], 'reformulate': [30], 'layers': [32, 85], 'as': [33], 'functions': [36], 'with': [37, 78, 130], 'reference': [38], 'layer': [41], 'i

---

The following code is updated so instead of slicing it takes use of the get function from the openalex API.

It returns each IC2S2 author, their ID and a coloumn containing relevant work information not including abstract which comes in a different coloumn.

In [10]:
import requests
import pandas as pd
from joblib import Parallel, delayed
from tqdm import tqdm
import time

def get_author(author_id):
    try:
        # Ensure author_id is a string
        author_id = str(author_id)

        # 1. Get basic author info
        author_url = f"https://api.openalex.org/authors/{author_id}"
        author_response = requests.get(author_url)
        if author_response.status_code != 200:
            return {"author_id": author_id, "error": "Author not found"}

        author_data = author_response.json()
        author_name = author_data.get("display_name", "Unknown")

        # 2. Get all author's works with pagination
        base_works_url = f"https://api.openalex.org/works?filter=author.id:{author_id}&per_page=200"
        all_works = []

        cursor = "*"
        while cursor:
            works_url = f"{base_works_url}&cursor={cursor}"
            works_response = requests.get(works_url)
            if works_response.status_code != 200:
                break
            works_data = works_response.json()
            works = works_data.get("results", [])
            all_works.extend(works)
            cursor = works_data.get("meta", {}).get("next_cursor", None)
            time.sleep(0.5)

        # Initialize data structure
        works_data = {
            "author_id": author_id,
            "author_name": author_name,
            "works": [],
            "abstract": []
        }

        # Process all collected works and check concepts
        for work in all_works:
            concepts = {c["display_name"] for c in work.get("concepts", []) if c["level"] == 0}
            social_science = {"Sociology", "Psychology", "Economics", "Political Science"}         
            quantitative_science = {"Mathematics", "Physics", "Computer Science"}
            if not (concepts & social_science) or not (concepts & quantitative_science):
                continue
            work_info = {
                "id": work.get("id"),
                "title": work.get("title"),
                "publication_year": work.get("publication_year"),
                "cited_by_count": work.get("cited_by_count"),
                "author_ids": [authorship.get("author", {}).get("id") 
                             for authorship in work.get("authorships", []) 
                             if authorship.get("author", {}).get("id")]
            }
            works_data["abstract"].append({"abstracts_inverted_index": work.get("abstract_inverted_index", "N/A")})
            works_data["works"].append(work_info)

        return works_data

    except Exception as e:
        return {"author_id": author_id, "error": str(e)}


We run the above function for all author IDs

In [11]:
# Load and clean DataFrame
df = pd.read_csv("authors_matched.csv")
df['ID'] = df['ID'].str.split('/').str[-1]
df = df[df["Works Count"] > 5]
df = df[df["Works Count"] < 5000]

# Run in parallel with progress bar
all_extracted_data = Parallel(n_jobs=5)(
    delayed(get_author)(author) for author in tqdm(df["ID"])
)

# Convert results to DataFrame (optional)
results_df = pd.DataFrame(all_extracted_data)


100%|██████████| 987/987 [11:55<00:00,  1.38it/s]


In [13]:
results_df

author_id              author_name  \
0    A5014394213            Xuechunzi Bai   
1    A5002575847            Angelina Wang   
2    A5053200092         Ilia Sucholutsky   
3    A5077079119      Thomas L. Griffiths   
4    A5030641600            Babak Heydari   
..           ...                      ...   
982  A5079825006         Thomas W. Malone   
983  A5108189679                   A. Wan   
984  A5047556152              Zhenguo Chi   
985  A5025218700          Amit Goldenberg   
986  A5058723753  Olivier Bergeron-Boutin   

                                                 works  \
0    [{'id': 'https://openalex.org/W3026440911', 't...   
1    [{'id': 'https://openalex.org/W4283166422', 't...   
2    [{'id': 'https://openalex.org/W4402512774', 't...   
3    [{'id': 'https://openalex.org/W2123713131', 't...   
4    [{'id': 'https://openalex.org/W3119614971', 't...   
..                                                 ...   
982  [{'id': 'https://openalex.org/W2128677857', 't...   
983  [{'id': 'https://openalex.org/W4404390391', 't...   
984  [{'id': 'https://openalex.org/W1996588871', 't...   
985  [{'id': 'https://openalex.org/W3137450077', 't...   
986  [{'id': 'https://openalex.org/W4200166430', 't...   

                                              abstract  
0    [{'abstracts_inverted_index': {'With': [0, 91]...  
1    [{'abstracts_inverted_index': {'Research': [0]...  
2    [{'abstracts_inverted_index': None}, {'abstrac...  
3    [{'abstracts_inverted_index': {'In': [0], 'com...  
4    [{'abstracts_inverted_index': {'Although': [0]...  
..                                                 ...  
982  [{'abstracts_inverted_index': {'This': [0], 's...  
983  [{'abstracts_inverted_index': {'How': [0], 'do...  
984  [{'abstracts_inverted_index': {'A': [0], 'nove...  
985  [{'abstracts_inverted_index': None}, {'abstrac...  
986  [{'abstracts_inverted_index': {'Abstract': [0]...  

[987 rows x 4 columns]

To get a DF for the works we use the following box

In [14]:
# Flatten the list of dictionaries into rows
df_expanded = results_df.explode(["works","abstract"])

# Convert dictionary values into separate columns
df_new = pd.json_normalize(df_expanded["works"])

# Add the corresponding abstract column
df_new['abstract'] = df_expanded['abstract'].values

# Display result
df_new

id  \
0      https://openalex.org/W3026440911   
1      https://openalex.org/W3092124192   
2      https://openalex.org/W4391632026   
3      https://openalex.org/W2979265655   
4      https://openalex.org/W4210705188   
...                                 ...   
15850  https://openalex.org/W4200166430   
15851  https://openalex.org/W4224998649   
15852  https://openalex.org/W4206700794   
15853  https://openalex.org/W4398411990   
15854  https://openalex.org/W4398639886   

                                                   title  publication_year  \
0      As diversity increases, people paradoxically p...            2020.0   
1      Comprehensive stereotype content dictionaries ...            2020.0   
2      Measuring Implicit Bias in Explicitly Unbiased...            2024.0   
3      Vertical and horizontal inequality are status ...            2019.0   
4      Face Stereotypes from Insufficient Statistical...            2021.0   
...                                                  ...               ...   
15850  The partisan nature of support for democratic ...            2021.0   
15851  COVID-19 and Support for Executive Aggrandizement            2022.0   
15852  Willful Blindness: Franco-American Relations a...            2019.0   
15853  Replication Data for: "Expert Bias and Democra...            2023.0   
15854  Replication Data for: Expert Bias and Democrat...            2023.0   

       cited_by_count                                         author_ids  \
0                72.0  [https://openalex.org/A5014394213, https://ope...   
1                69.0  [https://openalex.org/A5021657086, https://ope...   
2                 6.0  [https://openalex.org/A5014394213, https://ope...   
3                25.0  [https://openalex.org/A5080791781, https://ope...   
4                 1.0  [https://openalex.org/A5014394213, https://ope...   
...               ...                                                ...   
15850            71.0  [https://openalex.org/A5039738213, https://ope...   
15851             5.0  [https://openalex.org/A5039738213, https://ope...   
15852             1.0                 [https://openalex.org/A5058723753]   
15853             0.0  [https://openalex.org/A5058723753, https://ope...   
15854             0.0  [https://openalex.org/A5058723753, https://ope...   

                                                abstract  
0      {'abstracts_inverted_index': {'With': [0, 91],...  
1      {'abstracts_inverted_index': {'Abstract': [0],...  
2      {'abstracts_inverted_index': {'Large': [0], 'l...  
3                     {'abstracts_inverted_index': None}  
4      {'abstracts_inverted_index': {'Social': [0], '...  
...                                                  ...  
15850  {'abstracts_inverted_index': {'Abstract': [0],...  
15851  {'abstracts_inverted_index': {'Abstract': [0],...  
15852  {'abstracts_inverted_index': {'The': [0], 'lon...  
15853  {'abstracts_inverted_index': {'This': [0], 're...  
15854  {'abstracts_inverted_index': {'This': [0], 're...  

[15855 rows x 6 columns]

Implementing the last filters

In [15]:
import ast

df_filtered = df_new.copy()

# Remove rows where 'author_ids' is NaN
df_filtered = df_filtered.dropna(subset=["author_ids"])

# Only keep works that have been cited >10 times
df_filtered = df_filtered[df_filtered["cited_by_count"] > 10]

# Only keep works with <10 authors
df_filtered = df_filtered[df_filtered["author_ids"].apply(lambda x: len(x) < 10)]
df_filtered = df_filtered.drop_duplicates(subset=["id"])
df_filtered

id  \
0      https://openalex.org/W3026440911   
1      https://openalex.org/W3092124192   
3      https://openalex.org/W2979265655   
6      https://openalex.org/W4283166422   
7      https://openalex.org/W4281291869   
...                                 ...   
15833  https://openalex.org/W4252047720   
15836  https://openalex.org/W4253739739   
15837  https://openalex.org/W3010856299   
15838  https://openalex.org/W4238827467   
15850  https://openalex.org/W4200166430   

                                                   title  publication_year  \
0      As diversity increases, people paradoxically p...            2020.0   
1      Comprehensive stereotype content dictionaries ...            2020.0   
3      Vertical and horizontal inequality are status ...            2019.0   
6      Towards Intersectionality in Machine Learning:...            2022.0   
7      REVISE: A Tool for Measuring and Mitigating Bi...            2022.0   
...                                                  ...               ...   
15833                          Digital Emotion Contagion            2019.0   
15836                                Collective Emotions            2017.0   
15837  The Quest for Hope: Disadvantaged Group Member...            2020.0   
15838             The Crowd Emotion Amplification Effect            2020.0   
15850  The partisan nature of support for democratic ...            2021.0   

       cited_by_count                                         author_ids  \
0                72.0  [https://openalex.org/A5014394213, https://ope...   
1                69.0  [https://openalex.org/A5021657086, https://ope...   
3                25.0  [https://openalex.org/A5080791781, https://ope...   
6                50.0  [https://openalex.org/A5002575847, https://ope...   
7                47.0  [https://openalex.org/A5002575847, https://ope...   
...               ...                                                ...   
15833            22.0  [https://openalex.org/A5025218700, https://ope...   
15836            13.0  [https://openalex.org/A5025218700, https://ope...   
15837            12.0  [https://openalex.org/A5040468552, https://ope...   
15838            11.0  [https://openalex.org/A5025218700, https://ope...   
15850            71.0  [https://openalex.org/A5039738213, https://ope...   

                                                abstract  
0      {'abstracts_inverted_index': {'With': [0, 91],...  
1      {'abstracts_inverted_index': {'Abstract': [0],...  
3                     {'abstracts_inverted_index': None}  
6      {'abstracts_inverted_index': {'Research': [0],...  
7                     {'abstracts_inverted_index': None}  
...                                                  ...  
15833  {'abstracts_inverted_index': {'People': [0], '...  
15836  {'abstracts_inverted_index': {'When': [0], 'an...  
15837  {'abstracts_inverted_index': {'Within': [0], '...  
15838  {'abstracts_inverted_index': {'How': [0], 'do'...  
15850  {'abstracts_inverted_index': {'Abstract': [0],...  

[5167 rows x 6 columns]

Saving to files

In [16]:
IC2S2_papers = df_filtered[["id","publication_year","cited_by_count","author_ids"]]
IC2S2_abstracts = df_filtered[["id","title","abstract"]]

IC2S2_papers.to_csv("IC2S2_papers.csv")
IC2S2_abstracts.to_csv("IC2S2_abstracts.csv")

Finding all co authors avoiding duplicates and IC2S2 authors

In [44]:
co_authors = df_filtered["author_ids"].explode("author_ids")
co_authors = co_authors.str.split('/').str[-1]
# co_authors = co_authors.unique()
co_authors = co_authors[~co_authors.isin(df["ID"])]
co_authors = co_authors.unique()
co_authors.shape

(9165,)

**Dataset summary. How many works are listed in your IC2S2 papers dataframe? How many unique researchers have co-authored these works?**

We managed to find 3916 works in our IC2S2 papers dataframe, with 7017 unique co-authors included here.

**Efficiency in code. Describe the strategies you implemented to make your code more efficient. How did your approach affect your code's execution time? (answer in max 150 words)**

To optimize the efficiency of our filtering, we used multi-processing with pandas operations. Pandas operations are heavily optimized for data filtering and written in C for efficiency. Multi-processing let's us use all the cores of our CPU, which makes running operations much faster (but computational power equal). Therefore we have optimized both aspects, both speed and efficiency.

**Filtering Criteria and Dataset Relevance Reflect on the rationale behind setting specific thresholds for the total number of works by an author, the citation count, the number of authors per work, and the relevance of works to specific fields. How do these filtering criteria contribute to the relevance of the dataset you compiled? Do you believe any aspects of Computational Social Science research might be underrepresented or overrepresented as a result of these choices? (answer in max 150 words)**

The filters created a sense of relevance in the works. The total number of works by an author ensured that the works most likely was of a certain quality.

citation count made sure the paper actually discovered something of relevance as many citation most likely means new information.

Number of authors per work ensured that the speakers at IC2S2 were a real part of the work.

The relevance of the concept ensured that we only kept works relevant for computational social science ensuring only topics we are interested in were represented.

We underrepresent new researches by the lower bound of total works that may have discovered great things!

---

## Part 4

**part 1**

In [56]:
import networkx as nx
import ast
from networkx.readwrite import json_graph
import json

In [57]:
papers_df = pd.read_csv("IC2S2_papers.csv")

# Convert string representation of lists into actual Python lists
papers_df["author_ids"] = papers_df["author_ids"].apply(lambda x: ast.literal_eval(x) if isinstance(x, str) else x)
papers_df

Unnamed: 0                                id  publication_year  \
0              0  https://openalex.org/W3026440911            2020.0   
1              1  https://openalex.org/W3092124192            2020.0   
2              3  https://openalex.org/W2979265655            2019.0   
3              6  https://openalex.org/W4283166422            2022.0   
4              7  https://openalex.org/W4281291869            2022.0   
...          ...                               ...               ...   
5162       15832  https://openalex.org/W4252047720            2019.0   
5163       15835  https://openalex.org/W4253739739            2017.0   
5164       15836  https://openalex.org/W3010856299            2020.0   
5165       15837  https://openalex.org/W4238827467            2020.0   
5166       15849  https://openalex.org/W4200166430            2021.0   

      cited_by_count                                         author_ids  
0               72.0  [https://openalex.org/A5014394213, https://ope...  
1               69.0  [https://openalex.org/A5021657086, https://ope...  
2               25.0  [https://openalex.org/A5080791781, https://ope...  
3               50.0  [https://openalex.org/A5002575847, https://ope...  
4               47.0  [https://openalex.org/A5002575847, https://ope...  
...              ...                                                ...  
5162            22.0  [https://openalex.org/A5025218700, https://ope...  
5163            13.0  [https://openalex.org/A5025218700, https://ope...  
5164            12.0  [https://openalex.org/A5040468552, https://ope...  
5165            11.0  [https://openalex.org/A5025218700, https://ope...  
5166            71.0  [https://openalex.org/A5039738213, https://ope...  

[5167 rows x 5 columns]

In [58]:
authors_df = pd.read_csv("authors_matched.csv")
authors_df = authors_df.drop_duplicates(subset="ID")
authors_df

Name                                ID  \
0               Xuechunzi Bai  https://openalex.org/A5014394213   
1               Angelina Wang  https://openalex.org/A5002575847   
2            Ilia Sucholutsky  https://openalex.org/A5053200092   
3         Thomas L. Griffiths  https://openalex.org/A5077079119   
4               Babak Heydari  https://openalex.org/A5030641600   
...                       ...                               ...   
1143          Amit Goldenberg  https://openalex.org/A5025218700   
1144       Federico Zimmerman  https://openalex.org/A5061723972   
1145         Brooke Perreault  https://openalex.org/A5092145709   
1146  Olivier Bergeron-Boutin  https://openalex.org/A5058723753   
1147           Saumya Bhadani  https://openalex.org/A5089504371   

                                          Works API URL  Works Count  H-Index  \
0     https://api.openalex.org/works?filter=author.i...           33       10   
1     https://api.openalex.org/works?filter=author.i...           40       12   
2     https://api.openalex.org/works?filter=author.i...           76       10   
3     https://api.openalex.org/works?filter=author.i...          765       96   
4     https://api.openalex.org/works?filter=author.i...           97       20   
...                                                 ...          ...      ...   
1143  https://api.openalex.org/works?filter=author.i...          139       26   
1144  https://api.openalex.org/works?filter=author.i...            1        1   
1145  https://api.openalex.org/works?filter=author.i...            4        1   
1146  https://api.openalex.org/works?filter=author.i...            7        3   
1147  https://api.openalex.org/works?filter=author.i...            5        5   

     Country Code  
0              US  
1              US  
2              US  
3              US  
4              MX  
...           ...  
1143           US  
1144           AR  
1145           US  
1146           US  
1147           US  

[1148 rows x 6 columns]

In [ ]:
from collections import defaultdict
from itertools import combinations

def create_edge_list(df, authors_column):
    collaborations = defaultdict(int)
    
    for authors in df[authors_column]:
        for author1, author2 in combinations(sorted(authors), 2):
            collaborations[(author1, author2)] += 1
    
    edge_list = [(author1, author2, count) for (author1, author2), count in collaborations.items()]
    return edge_list

edge_list = create_edge_list(papers_df,'author_ids')
edge_list

[('https://openalex.org/A5014394213', 'https://openalex.org/A5080791781', 3),
 ('https://openalex.org/A5014394213', 'https://openalex.org/A5101722692', 1),
 ('https://openalex.org/A5080791781', 'https://openalex.org/A5101722692', 1),
 ('https://openalex.org/A5014394213', 'https://openalex.org/A5021657086', 1),
 ('https://openalex.org/A5021657086', 'https://openalex.org/A5080791781', 2),
 ('https://openalex.org/A5002575847', 'https://openalex.org/A5022811687', 2),
 ('https://openalex.org/A5002575847', 'https://openalex.org/A5073539430', 1),
 ('https://openalex.org/A5022811687', 'https://openalex.org/A5073539430', 1),
 ('https://openalex.org/A5001277622', 'https://openalex.org/A5002575847', 1),
 ('https://openalex.org/A5001277622', 'https://openalex.org/A5022811687', 1),
 ('https://openalex.org/A5001277622', 'https://openalex.org/A5025881129', 1),
 ('https://openalex.org/A5001277622', 'https://openalex.org/A5037438806', 1),
 ('https://openalex.org/A5001277622', 'https://openalex.org/A504

In [60]:
def create_graph(G,edge_list):
    for author1, author2, _ in edge_list:
        G.add_edge(author1,author2)
    return G

def add_weights(G,edge_list):
    G.add_weighted_edges_from(edge_list)
    return G

G = nx.Graph()
G = create_graph(G,edge_list)
G = add_weights(G,edge_list)
print(G)

Graph with 9791 nodes and 27887 edges


In [ ]:
def node_attributes(G, authors_df, papers_df):
    authors_info = authors_df.set_index('ID')[['Name', 'Country Code']].to_dict(orient='index')
    
    publication_info = {}

    for idx, row in papers_df.iterrows():
        author_list = row["author_ids"]
       
        pub_year = row['publication_year']
        cited_by = row['cited_by_count']
        
        for author in author_list:
            if author not in publication_info:
                publication_info[author] = {
                    'first_publication_year': pub_year,
                    'citation_count': cited_by
                }
            else:
                if pub_year < publication_info[author]['first_publication_year']:
                    publication_info[author]['first_publication_year'] = pub_year
                publication_info[author]['citation_count'] += cited_by
    
    for author in G.nodes():
        author_attr = authors_info.get(author, {})
        display_name = author_attr.get('Name')
        country = author_attr.get('Country Code')
        
        pub_info = publication_info.get(author, {})
        first_year = pub_info.get('first_publication_year')
        citation_count = pub_info.get('citation_count')
        
        combined_attr = {
            'display_name': display_name,
            'country': country,
            'first_publication_year': first_year,
            'citation_count': citation_count
        }
        G.nodes[author].update(combined_attr)
    
    return G

In [62]:
G = node_attributes(G,authors_df,papers_df)

In [ ]:
def save_graph_as_json(G, filename):
    data = json_graph.node_link_data(G)

    with open(filename, 'w') as f:
        json.dump(data, f, indent=4)

In [64]:
save_graph_as_json(G,"graph.json")

**part 2**

In [66]:
print(G)

Graph with 9791 nodes and 27887 edges


**What is the total number of nodes (authors) and links (collaborations) in the network?**

The total number of nodes for our graph is 9791 and for collaborations we have 27887.

**Calculate the network's density (the ratio of actual links to the maximum possible number of links). Would you say that the network is sparse? Justify your answer.**




In [70]:
2 * len(G.edges)/ (len(G.nodes())*(len(G.nodes())-1))

0.0005818647526980908

Given the low density the network it would seem quite sparse. It shows that not a lot of the combinations of different authors actually appears.

The authors are from all over the world so them not having worked together in all combinations is not suprising so the result makes sense.


**Is the network fully connected (i.e., is there a direct or indirect path between every pair of nodes within the network), or is it disconnected?**

The network is not fully connected, it is disconnected... Not everyone has written with someone that has written with someone else, etc. Some people have only written with a few others, and they are all writing with the same people in a smaller group (a group could be professors at a given uni).

**If the network is disconnected, how many connected components does it have? A connected component is defined as a subset of nodes within the network where a path exists between any pair of nodes in that subset.**



In [ ]:
components = list(nx.connected_components(G))

print(len(components))

287


We therefore have 287 connected components in our graph!

**How many isolated nodes are there in your network? An isolated node is defined as a node with no connections to any other node in the network.**



In [ ]:
isolated_nodes = list(nx.isolates(G))

print(len(isolated_nodes))

0


No isolated nodes therefore appear in our graph!

**Discuss the results above on network density, and connectivity. Are your findings in line with what you expected? Why? (answer in max 150 words)**

Yes it is in line with our expectations. The authors arent largely connected since they are from different fields, different countries and different universities. It would be very surprising if there was some connection to make between everyone. 

Furthermore the lack of any isolated nodes makes sense since we don't believe there to be any research being conducted completely alone with no co-authors. Research is often, at least, supervised or read through by others.